packages/wallet/src/redux/protocols/indirect-funding/player-b/reducer.ts

```diff
@@ -1,10 +1,219 @@
-import * as states from './state';
+import * as states from '../state';
+import { PlayerBState } from '../state';
+
+import * as actions from '../../../actions';
+
 import { ProtocolStateWithSharedData } from '../../';
-import { SharedData } from '../../../state';
+import { SharedData, checkAndStore, getChannel, signAndStore, queueMessage } from '../../../state';
+import { IndirectFundingState, failure, success } from '../state';
+import { unreachable } from '../../../../utils/reducer-utils';
+import {
+  BWaitForPreFundSetup0,
+  BWaitForDirectFunding,
+  BWaitForLedgerUpdate0,
+  BWaitForPostFundSetup0,
+  bWaitForDirectFunding,
+  bWaitForLedgerUpdate0,
+  bWaitForPostFundSetup0,
+} from './state';
+import { getChannelId, nextSetupCommitment } from '../../../../domain';
+import { CONSENSUS_LIBRARY_ADDRESS } from '../../../../constants';
+import { createCommitmentMessageRelay } from '../../reducer-helpers';
+import { theirAddress } from '../../../../redux/channel-store';
+import { initialDirectFundingState } from '../../direct-funding/state';
+
+import { directFundingRequested } from '../../direct-funding/actions';
+import { DirectFundingAction } from '../../direct-funding';
+import { directFundingStateReducer } from '../../direct-funding/reducer';
+import { isSuccess, isFailure } from '../../direct-funding/state';
+
+type ReturnVal = ProtocolStateWithSharedData<IndirectFundingState>;
+type IDFAction = actions.indirectFunding.Action;
+
+export function initialize(channelId: string, sharedData: SharedData): ReturnVal {
+  // todo: check that channel exists?
+  return { protocolState: states.bWaitForPreFundSetup0({ channelId }), sharedData };
+}
+
+export function playerBReducer(
+  protocolState: PlayerBState,
+  sharedData: SharedData,
+  action: IDFAction | DirectFundingAction,
+): ReturnVal {
+  switch (protocolState.type) {
+    case 'BWaitForPreFundSetup0':
+      return handleWaitForPreFundSetup(protocolState, sharedData, action);
+    case 'BWaitForDirectFunding': // defer to child reducer
+      return handleWaitForDirectFunding(protocolState, sharedData, action);
+    case 'BWaitForLedgerUpdate0':
+      return handleWaitForLedgerUpdate(protocolState, sharedData, action);
+    case 'BWaitForPostFundSetup0':
+      return handleWaitForPostFundSetup(protocolState, sharedData, action);
+    default:
+      return unreachable(protocolState);
+  }
+}
+
+function handleWaitForPreFundSetup(
+  protocolState: BWaitForPreFundSetup0,
+  sharedData: SharedData,
+  action: IDFAction | DirectFundingAction,
+): ReturnVal {
+  const unchangedState = { protocolState, sharedData };
+  if (action.type !== actions.COMMITMENT_RECEIVED) {
+    throw new Error('Incorrect action');
+  }
+
+  const checkResult = checkAndStore(sharedData, action.signedCommitment);
+  if (!checkResult.isSuccess) {
+    throw new Error('Indirect funding protocol, unable to validate or store commitment');
+  }
+  sharedData = checkResult.store;
```

Comment:
Isn't this mutating the existing state instead of creating a new copy (which can create a lot of headaches)?

Reply:
Nevermind, that doesn't seem to be the case.